### ISSUE TITLE
California Games not advancing past BMX event

### ISSUE THREAD

**Issue by sa666666** [stella-emu/stella#858]
Bug has apparently been present since Stella 3.9.3, and is still present in master!

ROM: [California Games (1987) (Epyx).zip](https://github.com/stella-emu/stella/files/7789606/California.Games.1987.Epyx.zip)
State file from 6.5.3 (submitted by bug reporter): [California Games (1987) (Epyx)_state_6.5.3.zip](https://github.com/stella-emu/stella/files/7789609/California.Games.1987.Epyx._state_6.5.3.zip)
State file from master: [California Games (1987) (Epyx)_state_6.7pre.zip](https://github.com/stella-emu/stella/files/7789670/California.Games.1987.Epyx._state_6.7pre.zip)

This was reported by PM to AtariAge; I will include the relevant info here.

------------------------
When playing California Games in normal mode (meaning that you're progressing from completing one event then moving on to the next), the game will not go past the BMX downhill event once complete.  It will just stay at the screen with the bike at the finish line and never move to the next screen.  It should display the team with the high score, then move on to the Surfing event.  I tried with other emulators and it behaves as expected, but when playing with Stella, it "hangs" after the BMX event.  Is this a known issue?

This save state0 will put you right before the last screen on the BMX event. Once you cross the finish line, you won’t be able to continue. It’s supposed to show the event winner screen and then go to the surfing event when you press the button.

I also added a save state for the end of the footbag event (state1).  You'll see that when you press the button, it behaves as it should for this event.

------------------------
I tried in Stella 6.5.3, and confirmed it there.  And also in latest master (as of Dec. 29), with the same results.  I set emulation speed to 1000%, to get to the desired point very quickly.

Some oddities:
 - The state save files provided for 6.5.3 are over 6MB each!  They should only be ~18-20KB.  But they still worked in 6.5.3.  Perhaps there's a truncation bug or something in the state save mechanism.  Maybe this should be split out as a separate issue.
 - If I play around with the emulation speed, it starts working!  So if I set the speed to, say, 25%, or 300%, pressing Space (aka, Fire button) allows the event to proceed.  But if I set it back to 100%, it doesn't work again :confused:

**Comment by thrust26** [comment]
Adding @DirtyHairy because the fact that it is working at **other** speeds than 100% makes me wondering.

**Comment by sa666666** [comment]
Yeah, that makes sense.

**Comment by thrust26** [comment]
The code (non intentionally?) also reads `INPT4` via `PLA`  multiple times during kernel display. I wonder if that affects Stella.

**Comment by thrust26** [comment]
Attached is a simple test program, which checks the fire button twice per frame just like CG does. The background color should change when the button is pressed or released. This does not happen in Stella but works for other emulators.

[test_button.zip](https://github.com/stella-emu/stella/files/7790405/test_button.zip)

**Comment by sa666666** [comment]
It doesn't work in MAME or z26 either, but does in Stellerator, gopher2600 and clksignal.

**Comment by thrust26** [comment]
Findings:
- my test ROM checks at scanlines 0 and 228 (CG at 3 and 221).
- on a real console this would mean, about 12.5% of state changes (between scanlines 229 and 262) would be missed 
- when testing with gopher2600 and Stellerator, it also misses a small fraction of state changes

Some thoughts:
- CG checks the fire button at the very beginning of a frame and after the display kernel
- the code expects an input state change somewhere in the middle of a frame
- in Stella all state changes happen between two frames
- if Stella would poll events **while** the `EmulationWorker` is running, CG would work. The 1st check would take the state from the previous poll and the 2nd from the current one.
- I suppose that's what the other, "good" emulators do 
- BUT: The faster the CPU and emulation, the less real time the `EmulationWorker` requires (theoretically zero). And then there will be no state changes possible during one frame.  
- so this only works by coincidence and only with slow enough emulation

Ideas for solutions:
1. We could "**slow down**" emulation, e.g. by emulating only fractions of a frame each step, poll events, sleep and then continue with the next step. Each step should take about the same amount of time. For CG two steps per frame would be sufficient, and I cannot think of any realistic scenario which would require more.
2. We could poll events parallel to the `EmulationWorker` and cross our fingers, hoping that the first check happens **before** we have polled new events.
3. We could move the beginning of a frame e.g. to the start of the display kernel. So the VBlank code reads the result from the current, and Overscan the result of the next poll.
4. ...?

**Comment by thrust26** [comment]
@DirtyHairy @sa666666 How shall we continue here?

**Comment by sa666666** [comment]
How about, in addition to the normal polling, we poll at each read of INPTx too?  Would this cause too much slowdown, I wonder?

If there are no pending events, this is essentially a small check (check returns false, so method returns immediately).

**Comment by thrust26** [comment]
We are doing this for paddle and trackball games anyway, no?

**Comment by sa666666** [comment]
No, currently all controllers are querying `myEvent`, which is essentially a cached state of the events from the last time polling was done in `EventHandler`.  What I am proposing is to directly call `EventHandler::pollEvent` at each INPTx read.  If there are no pending events from SDL, then this is a simple check (`while(SDL_PollEvent(&myEvent))`) that exits immediately.  Otherwise it will get the pending events and put them in `myEvent`, from which the current code can then read it and continue as it is now.

**Comment by thrust26** [comment]
Hm, paddles are read multiple times per frame. So here Stella must update multiple times per frame. We should do the same with joystick controllers.

**Comment by sa666666** [comment]
Well, first I would try it on every read, to see if it even fixes the issue.  Then if that works, we can further optimize it.

**Comment by sa666666** [comment]
It's going to require some refactoring.  New method(s) need to be added to `EventHandler`, and then the object needs to be accessible from the Controller class.  I suggest whatever we do, get it test "quick and dirty" first, to see if it will even help.  Then we can worry about doing it cleanly.

**Comment by thrust26** [comment]
Wouldn't a change in `OSystem::mainLoop()` be less invasive?

**Comment by sa666666** [comment]
I don't know.  And I guess we need to experiment.  I'm not even entirely sure my suggestion will fix the problem.  I think this work needs to be done in a branch first, since it has the potential to be very invasive either way.

**Comment by DirtyHairy** [comment]
Be careful with this. Afaik accessing input is only legal on the main thread on some platforms, and emulation runs on thread.

**Comment by thrust26** [comment]
How about a separate thread which only polls the input multiple times per second?

**Comment by thrust26** [comment]
I experimented with adding intermediate polls and sleeps into `OSystem::mainLoop()`. I can see they are getting executed, but this had no effect. 😕 What's wrong here?

```
    time_point<high_resolution_clock> oldVirtualTime = virtualTime;

    virtualTime += duration_cast<high_resolution_clock::duration>(timeslice);
    ...
    else if (virtualTime > now) {

      constexpr int polls_per_frame = 16;
      for(int i = 1; i < polls_per_frame; ++i)
      {
        time_point<high_resolution_clock> sleepVirtualTime = oldVirtualTime +
          duration_cast<high_resolution_clock::duration>(timeslice * i / polls_per_frame);
        if(sleepVirtualTime > now)
        {
          cerr << i << " ";
          std::this_thread::sleep_until(sleepVirtualTime);
          myEventHandler->poll(TimerManager::getTicks());
        }
      }
      cerr << endl;
   
      // Wait until we have caught up with 6507 time
      std::this_thread::sleep_until(virtualTime);
    }
      
```

**Comment by thrust26** [comment]
> Be careful with this. Afaik accessing input is only legal on the main thread on some platforms, and emulation runs on thread.

Two ideas here:

1. interrupt emulation when INPT4/5 are accessed, sleep*, poll and continue emulation
2. sleep* emulation when INPT4/5 are accessed, poll multiple times per frame in a separate thread

* sleep until current real time

**Comment by thrust26** [comment]
@DirtyHairy @sa666666 How should we proceed here?

**Comment by thrust26** [comment]
@DirtyHairy @sa666666 Bump. IMO we should fix this for 7.0.

**Comment by thrust26** [comment]
@DirtyHairy @sa666666 2nd Bump

**Comment by jdruding** [comment]
I'm not sure if there is anything I can do to help with this issue, but I do have some technical skills and I wanted to over to at least offer to help test as needed.   California Games is one of my childhood favorites so I'm willing to help in any way I can! :-)

**Comment by thrust26** [comment]
Testing is not our major concern here. We have identified the problem and know how to test it. But fixing it is quite a bit more difficult.

**Comment by sa666666** [comment]
I recently worked on adding proper PA7 emulation in #108, and fixed it in https://github.com/stella-emu/stella/commit/114d169a46c3d4e88b3d97f872f4385ea6a7d4cf.  In doing this, I had to check the state much more frequently, essentially on every peek/poke, rather than once per frame as before.  I suspect the same will be required here.

**Comment by thrust26** [comment]
When we analyzed this, we found that the game checks twice per frame (before and after the kernel).

**Comment by sa666666** [comment]
OK, and I'm working on a way to tag events to deal with this.  Still WIP; I'll report back when I have more info.

**Comment by sa666666** [comment]
@thrust26 and @DirtyHairy 

I have this working in `feature/subframe-poll`.  But as usual, I'll need feedback on whether this fixes the actual issue, or just papers over it.  I think this one actually fixes it, since it's not a change to the emulation core itself, but to how it gathers and disperses events.

**Comment by thrust26** [comment]
So you update the controller state before you read SWCHA/SWCHB, correct? IIRC we had discussed this as a valid solution. 

BTW: Maybe the trackball is missing, but this must have been updated very frequently already, Is that a different mechanism. Which should be aligned?

**Comment by sa666666** [comment]
If you're referring to https://github.com/stella-emu/stella/issues/689, that is addressed here as well.  This code passes all test ROMs from there, and all the ones here as well.

Attached is a state file for the ROM above.  It's at the end of the BMX event.  I suggest compiling two versions of Stella; one from `master` and one from `feature/subframe-poll`.  After loading the state file, for the former, pressing Fire doesn't advance the player; for the latter it does.

[California Games (1987) (Epyx).zip](https://github.com/user-attachments/files/29003567/California.Games.1987.Epyx.zip)

**Comment by sa666666** [comment]
We can all discuss the specifics of the code if you like.  Although the comments explain it pretty well, I can go into the overall big picture.

**Comment by DirtyHairy** [comment]
Read over the code. You're essentially "smearing" all events that happened between two consecutive frames over the second frame uniformly. This is a valid approach. I have a few remarks

* I would use a `uInt32` instead of a `float`.
* Mouse buttons (i.e. on `Joystick.cxx` do not replay the recorded events, but report the aggregated value instead. Is there a specific reason why you are not binding them to their respective events?
* Autofire is another special case, and there is an ugly special case for it in various places. I think it would be better to introduce an "autofire" event, bind to that if autofire is enabled and let the event layer handle the autofire sequence
* I'd strongly suggest to replace the explicit `myHasTransitions.load` / `myHasTransitions.store` with plain `operator=` / `operator bool()` --- that's equivalent to an implicit `memory_order_seq_cst` and **much** easier to reason about than `memory_order_relaxed`. The slightly more surface for contention, but totally irrelevant for us. As an alternative, keep the `load` / `store`, but remove the explicit memory ordering.

**Comment by sa666666** [comment]
OK, I'll look into those options and report back later.

**Comment by sa666666** [comment]
I have to think about this a little more.  I'm still not entirely comfortable with the timing being scanline-based, and think maybe it should be cycle-based.  After all, a real controller and the RIOT have no knowledge of the TIA or scanlines, so maybe our emulation shouldn't either.

**Comment by DirtyHairy** [comment]
Yeah, I had the same thought, but I couldn't come up with a better scheme. Our timeslices are not uniform. **Usually** they consist of a single frame, but apart from the fact that they are bounded by a min / max number of cycles you don't know much about them in advance.

Actually, thinking of it, the current approach has issues. If a ROM produces a long frame and then a short frame (which can happen for a malformed TV signal), then there is a part of the position interval that is not accessible during the short frame, and they will be subsequently discarded. To fix this, you'll have to keep unprocessed events in `beginInputWindow`.

In addition, there may be more than one timeslice between polling the SDL event loop (for example if the rendering takes too long), and this currently means that the scanline will wrap back to zero and the events will be replayed again. So, you have to actually keep track of the last "position" and ignore requests to read below it.

If you switch to cycles, distribute your events over the expected number of cycles for the next timeslice (a single frame) and move events that have not been consumed to the next timeslice, then cycles would work out, too. However, I don't see an advantage over scanlines, as the assignment is just as artificial.

But I think those two issues need to be fixed anyway (replaying events when the line wraps, and dropping events it the next slice is shorter than expected).

**Comment by sa666666** [comment]
@DirtyHairy, I think all items you mentioned are now done, except for the autofire as an event.  We can come back to that later, as it has nothing to do with this issue.

**Comment by thrust26** [comment]
I still have the question regarding the trackball. Wouldn't the new solution work for them too? So why are the classes not touched?

**Comment by sa666666** [comment]
@thrust26, are you talking about the `7800test_OK.bin` and `7800test_NOK.bin` ROMs?  If so, these are fixed because the old code was using `TIA::scanlines()` as its timing source, and the new code is using `System::cycles()`.  And the required changes to fix them are happening at a lower level, directly in `Event` and `Controller` classes.

If you're talking about something else, then please elaborate.

**Comment by thrust26** [comment]
@sa666666 No, I am talking about controller handling in general and the changes you made. The trackball frequently reads SWCHA during each frame (maybe 50x). So it acts like CG BMX on steroids. 😄 Yet, its classes are not affected by your changes, What am I missing here?

**Comment by sa666666** [comment]
What ROMs should I be testing for this?

**Comment by sa666666** [comment]
@thrust26, let's make sure we're on the same page here.  Are you saying that there are trackball (aka, `PointingDevice`) controllers  that aren't working as they should?  If so, which ones?  Or are you asking about how this refactor works in general?

If the latter, then I've changed nothing in `PointingDevice`-derived controllers, since the fix is at a lower level, in `Event` and `Controller`.  Basically, as @DirtyHairy mentioned above, the events are read in and 'smeared' over the next frame(s).  They're timestamped and applied later.  If 5 events occur at different times, each is sent to 1/5 of the frame (ie, the 'smearing').  This has the effect of spreading the events over the entire frame, and ROMs that do multiple reads per frame are then more likely to see them.  Nothing changed in the `PointingDevice` controller itself, since from the controller perspective it's still polling SWCHA as normal.  What *has* changed is that now there are (hopefully) more accurate events to read when the SWCHA poll *does*  happen.

**Comment by DirtyHairy** [comment]
> @DirtyHairy, I think all items you mentioned are now done, except for the autofire as an event. We can come back to that later, as it has nothing to do with this issue.

Looks good to me. However, I'd add a view more newlines in `Event.hxx`, and the comments and naming in there are pretty much misleading --- imo it shouldn't care about frames at all, but rather about input windows. Generally speaking, I find the comments left by Claude much too chatty and verbose, but that's a matter of taste. Also, it is worrying too much about premature optimisation ("we avoid the mutex-locked transition scan") --- performance here is entirely non-critical.

**Comment by sa666666** [comment]
Thanks, I'll address those items soon.

**Comment by sa666666** [comment]
OK, latest commit should take care of it.  Tightened up the comments, and changed references from frame/sub-frame to window/input window.

**Comment by thrust26** [comment]
> [@thrust26](https://github.com/thrust26), let's make sure we're on the same page here. Are you saying that there are trackball (aka, `PointingDevice`) controllers that aren't working as they should? If so, which ones? Or are you asking about how this refactor works in general?

The latter.

> If the latter, then I've changed nothing in `PointingDevice`-derived controllers, since the fix is at a lower level, in `Event` and `Controller`. Basically, as [@DirtyHairy](https://github.com/DirtyHairy) mentioned above, the events are read in and 'smeared' over the next frame(s). They're timestamped and applied later. If 5 events occur at different times, each is sent to 1/5 of the frame (ie, the 'smearing'). This has the effect of spreading the events over the entire frame, and ROMs that do multiple reads per frame are then more likely to see them. Nothing changed in the `PointingDevice` controller itself, since from the controller perspective it's still polling SWCHA as normal. What _has_ changed is that now there are (hopefully) more accurate events to read when the SWCHA poll _does_ happen.

So why did you change many other controller codes then?

**Comment by sa666666** [comment]
OK, now I understand what you're asking.

The changes I (and Claude) made are for digital I/O.  Specifically to fix the case where CalGames failed, which is because it polls SWCHA more than once per frame.  The code essentially timestamps an input, and plays it out over the next input window (aka, a frame most of the time, but sometimes multiple frames).  We had to do this, since there's no other way to get input data sampled fast enough and applied at the approx. right time.

Controllers with digital inputs (buttons for now, but can be extended to directions later too) have some updated code to deal with this 'replaying'.  The bulk of it is done in `Event.hxx` itself, but some small changes were needed for some of the controllers.  So the entire model is about transitions and an 'evenly spaced' replay of them.

`PointingDevice` controllers are different.  For their `::read()` method, it directly changes position based on cycle counts, so every SWCHA read is actually seeing different, updated output.  In some sense, it's even more accurate than the transition approach.  Because it's already essentially keeping track of multiple changes per frame, nothing had to change there.  The code I'm about to commit has this, which probably explains it a little better:
```
uInt8 PointingDevice::read()
{
  // SWCHA may be polled many times per frame, so the gray-code pins are sampled
  // cycle-accurately on every read -- this controller is sub-window accurate by
  // construction.  It deliberately does NOT use the Event transition schedule
  // (which replays discrete user-input events): the encoder instead synthesizes
  // an evenly-spaced pulse train from the whole-window mouse delta (see update())

  // Elapsed CPU cycles since the start of the current input window; this is
  // the controller's only notion of time, just as a real quadrature encoder
  // emits transitions purely as a function of elapsed time
  const int elapsed = static_cast<int>(mySystem.cycles() - myWindowStartCycle);
```
Hopefully that makes more sense.  The one thing that **did** change in `PointingDevice`, and in all other Controllers too, is that the idea of TIA, scanlines, and frames has been completely removed.  That was a holdover from the z26 code originally used to model this.  Now everything is based on timing/cycles.

This actually also fixes https://github.com/stella-emu/stella/issues/689 too, since it's essentially the same problem.  Too frequent port reads wasn't working because the input wasn't there when SWCHA looked for it.  Now it is.

Hopefully this makes more sense.  If you and @DirtyHairy are clear and agree, I will commit this soon.

**Comment by thrust26** [comment]
OK, I now get why you did no change the axis part of the trackballs. And I suppose the trackball's fire button is handled by other changes you made?

**Comment by sa666666** [comment]
The trackball buttons are not currently handled by the new event model; I can add them if you think it's necessary.  I just thought that there are not many trackball games, and nobody mentioned issues with multiple reads per frame for them, so I left that part out.

**Comment by sa666666** [comment]
@thrust26, after thinking about it, I just added trackball button(s) to the new event model as well.

**Comment by thrust26** [comment]
While you're at it, wouldn't it make sense to do the same for controller directions? I know of no other game besides CG BMX that does this, but I think it would enhance emulation and provide consistency.

**Comment by sa666666** [comment]
It's already done for controller directions for joystick, and all the controllers that inherit from it.  It doesn't make sense to do it for `PointingDevice` directions, as we just discussed a few posts ago.

This is the current status:
```
  Cycle-based sub-frame input (#858) — current state

  Core work: DONE + VALIDATED. The whole #858 sub-frame input model was moved off TIA scanlines onto System::cycles(),
  decoupling the RIOT and all controllers from the TIA (faithful to real hardware — controllers/RIOT share only the
  oscillator clock, not TIA state). Event owns the per-window cycle snapshot; M6532 makes zero tia() calls now. Branch:
  feature/subframe-poll, builds clean (warnings-as-errors, plain make).

  Latest session (2026-06-18), two things landed:
  - PointingDevice fire button → Event model (committed 74621ee9a): the trackball's pin Six now bindPins + replays sub-window
  via the shared updateFireButton helper (motion stays as cycle-accurate gray-code synth, deliberately not on the Event
  schedule). Button update moved out of the !myMouseEnabled gate so LeftJoystickFire drives it unconditionally.
  Build-verified only — not yet physically tested.
  - Conversion scope now COMPLETE except Lightgun (full controller survey, Stephen confirmed): Joystick family, Driving,
  Paddles, Switches, PointingDevice all converted. Everything else (MindLink, KidVid, AtariVox/SaveKey, Keyboard, CompuMate,
  QuadTari pass-through, analog pots) deliberately not converted — recorded so they're not re-proposed. 
```

**Comment by thrust26** [comment]
Totally fine by me now.

**Comment by sa666666** [comment]
OK, @DirtyHairy, if you're good with this too, I can commit it??

**Comment by sa666666** [comment]
@thrust26 is fine with this, and as I've addressed @DirtyHairy comments, I will commit this.

**Comment by sa666666** [comment]
Fixed in 65fc7d66abfc8dfba5033d88523437b83664ea2b.

**Comment by DirtyHairy** [comment]
Sorry for the late reply. Yes, I am fine with the changeset.
